<br><br>

---

# MODULE 01 - HOMEWORK: INTRO

---

<br><br>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

<br><br>

---

## Q1. Pandas version

What version of Pandas did you install?

In [2]:
pd.__version__

'3.0.6'

<br><br>

---
---

## Getting the data

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv')

In [4]:
df.head()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


<br><br>

---

## Q2. Records count

How many records are in the dataset?

---

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   model_year           10000 non-null  int64  
 1   origin               10000 non-null  str    
 2   fuel_type            10000 non-null  str    
 3   drivetrain           10000 non-null  str    
 4   num_doors            10000 non-null  int64  
 5   engine_displacement  10000 non-null  int64  
 6   num_cylinders        10000 non-null  int64  
 7   horsepower           9123 non-null   float64
 8   vehicle_weight       10000 non-null  int64  
 9   acceleration         9736 non-null   float64
 10  fuel_efficiency_mpg  10000 non-null  float64
dtypes: float64(3), int64(5), str(3)
memory usage: 859.5 KB


<br>

<span style='color:blue'>
    
**Answer:** 10000 records

<br>

<br><br>

---

## Q3. Fuel Types


How many fuel types are presented in the dataset?

---

In [6]:
df.fuel_type.unique()

<StringArray>
['Gasoline', 'Diesel', 'Hybrid']
Length: 3, dtype: str

<br>

<span style='color:blue'>
    
**Answer:** 3 Types (`Gasoline`, `Diesel`, and `Hybrid`)

<br>

<br><br>

---

## Q4. Missing Values


How many columns in the dataset have missing values?

---

In [7]:
df.isna().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

<br>

<span style='color:blue'>
    
**Answer:** There are 2 columns with missing values (`horsepower`, and `acceleration`)

<br>

<br><br>

---

## Q5. Max fuel efficiency


What's the maximum fuel efficiency of cars from Asia?

---

In [8]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

np.float64(41.2)

<br>

<span style='color:blue'>
    
**Answer:** 41.2

<br>

<br><br>

---

## Q6. Median value of horsepower


>1. Find the median value of the horsepower column in the dataset.
>
>2. Next, calculate the most frequent value of the same horsepower column.
>
>3. Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
>
>4. Now, calculate the median value of horsepower once again.

---

In [9]:
# Find the median value of the horsepower column in the dataset.
median_val = df.horsepower.median()
median_val

np.float64(254.0)

In [10]:
#  calculate the most frequent value of the same horsepower column.
frequent_val = df.horsepower.mode()[0]
frequent_val

np.float64(252.0)

In [11]:
# Use the fillna method to fill the missing values in the horsepower column with the most frequent value.
values = {"horsepower": frequent_val}
df = df.fillna(value=values)

df.isna().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower               0
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

In [12]:
# Now, calculate the median value of horsepower once again.
new_median_val = df.horsepower.median()
new_median_val

np.float64(252.0)

<br>

<span style='color:blue'>
    
**Has it Changed?** Yes, it decreased from 254 to 252

<br>

<br><br>

---

## Q7. Sum of weights


>1. Select all the cars from Asia
>
>2. Select only columns `vehicle_weight` and `model_year`
>
>3. Select the first 7 values
>
>4. Get the underlying NumPy array. Let's call it `X`.
>
>5. Compute matrix-matrix multiplication between the `transpose of X` and `X`. To get the transpose, use X.T. Let's call the result `XTX`.
>
>6. Invert XTX.
>
>7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
>
>8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
>
>9. What's the sum of all the elements of the result?

---

In [13]:
# 1. Select all the cars from Asia
# 2. Select only columns vehicle_weight and model_year
df_filt = df.query("origin == 'Asia'")[['vehicle_weight', 'model_year']]

# 3.Select the first 7 values
reduced_df = df_filt.iloc[:7]
reduced_df

,vehicle_weight,model_year
2,4240,1996
6,4310,1992
8,4230,1997
10,4420,1978
15,3950,2007
16,4750,2021
22,4450,2019


In [14]:
# 4. Get the underlying NumPy array. Let's call it X.
X = np.array(reduced_df)
X

array([[4240, 1996],
       [4310, 1992],
       [4230, 1997],
       [4420, 1978],
       [3950, 2007],
       [4750, 2021],
       [4450, 2019]])

In [15]:
# 5.Compute matrix-matrix multiplication between the transpose of X and X. 

XTX = X.T @ X
XTX

array([[131950500,  60750580],
       [ 60750580,  28041424]])

In [16]:
# 6.Invert XTX.
XTX_inverted = np.linalg.inv(XTX)
XTX_inverted

array([[ 2.96830537e-06, -6.43071025e-06],
       [-6.43071025e-06,  1.39675281e-05]])

In [17]:
# 7.Create an array "y" with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

In [18]:
# 8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w = XTX_inverted @ X.T @ y
w

array([0.13644777, 0.2327492 ])

In [19]:
# 9.What's the sum of all the elements of the result?
result = w.sum()
result

np.float64(0.36919696904925486)